File: Sudoku_Solver.java
public class Solution {
    public void solveSudoku(char[][] board) {
        if(board == null || board.length != 9 || board[0].length != 9) return;  
        backtrack(board, 0, 0);
    }
    
    private boolean backtrack(char[][] board, int x, int y) {
        if(x == 9) return true;
        if(y == 9) return backtrack(board, x + 1, 0);
        
        if(board[x][y] == '.') {
            for(char c = '1'; c<='9'; c++) {
                board[x][y] = c;
                if(isValid(board, x, y) && backtrack(board, x, y + 1)) return true;
                board[x][y] = '.';
            }
            return false;
        }
        else {
            return backtrack(board, x, y+1);
        }
    }
    
    private boolean isValid(char[][] board, int i, int j) {  
        for(int k=0;k<9;k++)  
        {  
            if(k!=j && board[i][k]==board[i][j])  
                return false;  
        }  
        for(int k=0;k<9;k++)  
        {  
            if(k!=i && board[k][j]==board[i][j])  
                return false;  
        }          
        for(int row = i/3*3; row<i/3*3+3; row++)  
        {  
            for(int col=j/3*3; col<j/3*3+3; col++)  
            {  
                if((row!=i || col!=j) && board[row][col]==board[i][j])  
                    return false;  
            }  
        }  
        return true;  
    }  
}

File: Swift/sudoku-solver.swift
/**
 * https://leetcode.com/problems/sudoku-solver/
 * 
 * 
 */ 
// Date: Mon Aug 30 09:48:46 PDT 2021
class Solution {
    func solveSudoku(_ board: inout [[Character]]) {
        var row = Array(repeating: Set(1...9), count: 9)
        var col = Array(repeating: Set(1...9), count: 9)
        var block = Array(repeating: Set(1...9), count: 9)

        func precompute() {
            for x in 0 ..< 9 {
                for y in 0 ..< 9 {
                    if let num = Int(String(board[x][y])) {
                        row[x].remove(num)
                        col[y].remove(num)
                        let blockIndex = (x / 3) * 3 + (y / 3)
                        block[blockIndex].remove(num)
                    }
                }
            }
        }

        @discardableResult func dfs(_ x: Int, _ y: Int) -> Bool {
            if x == 9 { return true }
            if y == 9 { return dfs(x + 1, 0) }
            if board[x][y] != Character(".") { return dfs(x, y + 1) }
            let candidates = row[x].intersection(col[y]).intersection(block[(x / 3) * 3 + (y / 3)])
            for candidate in candidates {
                row[x].remove(candidate)
                col[y].remove(candidate)
                block[(x / 3) * 3 + (y / 3)].remove(candidate)
                board[x][y] = Character(String(candidate))
                if dfs(x, y + 1) { return true }
                board[x][y] = Character(String("."))
                row[x].insert(candidate)
                col[y].insert(candidate)
                block[(x / 3) * 3 + (y / 3)].insert(candidate)
            }
            return false
        }

        precompute()
        dfs(0, 0)
    }
}